tests/frontier/scenarios/common.py

```diff
@@ -0,0 +1,302 @@
+        if res.result == ScenarioExpectOpcode.CALL_DATALOAD_0:
+            return env.call_dataload_0
+        if res.result == ScenarioExpectOpcode.CALL_DATASIZE:
+            return env.call_datasize
+        if res.result == ScenarioExpectOpcode.GASPRICE:
+            return exec_env.gasprice
+        if res.result == ScenarioExpectOpcode.COINBASE:
+            return int(exec_env.coinbase.hex(), 16)
+        if res.result == ScenarioExpectOpcode.TIMESTAMP:
+            return exec_env.timestamp
+        if res.result == ScenarioExpectOpcode.NUMBER:
+            return exec_env.number
+        if res.result == ScenarioExpectOpcode.GASLIMIT:
+            return exec_env.gaslimit
+        if res.result == ScenarioExpectOpcode.SELFBALANCE:
+            return int(env.selfbalance)
+    else:
+        return res.result
+    return 0
+
+
+def replace_special_calls_in_operation(
+    pre: Alloc, fork: Fork, operation: Bytecode, external_address: Address
+) -> Bytecode:
+    """
+    Run find replace of some special calls to the contracts that we don't know at compile time
+    replace 0xfff..fff address to external_address
+    replace special call to 0xfff..ffe address to gas_hash_address contract.
+    """
+    gas_hash_address = make_gas_hash_contract(pre)
+    invalid_opcode_contract = make_invalid_opcode_contract(pre, fork)
+
+    replace_list: List[Tuple[str, str]] = [
+        (
+            "602060646020604060007f"
+            + hex(SpecialAddress.INVALID_CALL_ADDRESS)[2:].lower()
+            + "611388f1",
+            Op.CALL(Op.SUB(Op.GAS, 200000), invalid_opcode_contract, 0, 64, 32, 100, 32).hex(),
+        ),
+        (
+            "600060006020604060007f" + hex(SpecialAddress.GAS_HASH_ADDRESS)[2:].lower() + "5af1",
+            Op.CALL(Op.SUB(Op.GAS, 200000), gas_hash_address, 0, 64, 32, 0, 0).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "31",
+            Op.BALANCE(external_address).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "3b",
+            Op.EXTCODESIZE(external_address).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "3f",
+            Op.EXTCODEHASH(external_address).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "3c",
+            "3c".join(Op.BALANCE(external_address).hex().rsplit("31", 1)),
+        ),
+    ]
+    for find, replace in replace_list:
+        if find in operation.hex():
+            new_operation_hex = operation.hex().replace(
+                find,
+                replace,
+            )
+            operation = Bytecode(
+                bytes_or_byte_code_base=to_bytes(new_operation_hex),
+                popped_stack_items=0,
+                pushed_stack_items=0,
+            )
+    return operation
+
+
+def make_gas_hash_contract(pre: Alloc) -> Address:
```

Comment:
This function can be a fixture, we should create `conftest.py` and put many of the functions there as fixtures.

Reply:
it is now a possible import for test code generators (ScenarioTestProgram class)